README.md
# noavoice
<!-- ALL-CONTRIBUTORS-BADGE:START - Do not remove or modify this section -->
[![All Contributors](https://img.shields.io/badge/all_contributors-4-orange.svg?style=flat-square)](#contributors-)
<!-- ALL-CONTRIBUTORS-BADGE:END -->

## 音声ファイルのLICENCEについて

以下の条件1〜3を満たしていれば、二次創作物の再配布を認めます。一次創作物の再配布は認めません。

1. **表示** — 適切な[クレジット](#クレジット)を表示し、[ライセンス](https://github.com/hrdtbs/noavoice/blob/master/README.md#%E9%9F%B3%E5%A3%B0%E3%83%95%E3%82%A1%E3%82%A4%E3%83%AB%E3%81%AElicence%E3%81%AB%E3%81%A4%E3%81%84%E3%81%A6)へのリンクを提供すること。

2. **非営利** — 営利目的で利用してはなりません。

3. **継承** — この作品の貢献した部分については同様のライセンスの下に頒布しなければなりません。

### クレジット

使用している音声が、**望月のあ**の音声であることが明確に分かるような形で、
**望月のあ**の名前とTwitterのID及びリンクを設置すること。

例

> 声 ： 望月のあ（[@_noach](https://twitter.com/_noach))

## 音声ファイル

上記、音声ファイルのLICENCEについてをちゃんとご覧になり遵守した上で[ここ](https://github.com/hrdtbs/noavoice/tree/master/public/static/voices
)からダウンロードを行い使用してください。


## Contributors ✨

Thanks goes to these wonderful people ([emoji key](https://allcontributors.org/docs/en/emoji-key)):

<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->
<!-- prettier-ignore-start -->
<!-- markdownlint-disable -->
<table>
  <tr>
    <td align="center"><a href="https://twitter.com/meowsmemo"><img src="https://avatars3.githubusercontent.com/u/30747709?v=4" width="100px;" alt=""/><br /><sub><b>hrdtbs</b></sub></a><br /><a href="https://github.com/hrdtbs/noavoice/commits?author=hrdtbs" title="Code">💻</a> <a href="https://github.com/hrdtbs/noavoice/commits?author=hrdtbs" title="Documentation">📖</a> <a href="#design-hrdtbs" title="Design">🎨</a> <a href="#ideas-hrdtbs" title="Ideas, Planning, & Feedback">🤔</a></td>
    <td align="center"><a href="https://ykzts.com/"><img src="https://avatars0.githubusercontent.com/u/12539?v=4" width="100px;" alt=""/><br /><sub><b>Yamagishi Kazutoshi</b></sub></a><br /><a href="https://github.com/hrdtbs/noavoice/commits?author=ykzts" title="Code">💻</a> <a href="#infra-ykzts" title="Infrastructure (Hosting, Build-Tools, etc)">🚇</a></td>
    <td align="center"><a href="https://himanoa.net"><img src="https://avatars2.githubusercontent.com/u/18651963?v=4" width="100px;" alt=""/><br /><sub><b>himanoa</b></sub></a><br /><a href="https://github.com/hrdtbs/noavoice/commits?author=himanoa" title="Code">💻</a> <a href="#infra-himanoa" title="Infrastructure (Hosting, Build-Tools, etc)">🚇</a><a href="#mochinoa-himanoa" title="Great fan of Mochiduki Noa">🐈🏵</a></td>
    <td align="center"><a href="http://saitoeku3.dev"><img src="https://avatars0.githubusercontent.com/u/38908416?v=4" width="100px;" alt=""/><br /><sub><b>Tadao Iseki</b></sub></a><br /><a href="https://github.com/hrdtbs/noavoice/commits?author=saitoeku3" title="Code">💻</a><a href="#design-hrdtbs" title="Design">🎨</a><a href="#mochinoa-saitoeku3" title="Great fan of Mochiduki Noa">🐈🏵</a></td>
  </tr>
</table>

<!-- markdownlint-enable -->
<!-- prettier-ignore-end -->
<!-- ALL-CONTRIBUTORS-LIST:END -->

This project follows the [all-contributors](https://github.com/all-contributors/all-contributors) specification. Contributions of any kind welcome!

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [2.2.1](https://github.com/hrdtbs/noavoice/compare/v2.2.0...v2.2.1) (2020-03-25)

## [2.2.0](https://github.com/hrdtbs/noavoice/compare/v2.1.0...v2.2.0) (2020-03-14)


### Features

* manage the current playing voice ([bc330ff](https://github.com/hrdtbs/noavoice/commit/bc330ff20c61b6d53e05c9f8ad7aa292ff049c23))
* Stop playing voice if exist playing voice ([846f0af](https://github.com/hrdtbs/noavoice/commit/846f0afebcaf29042df4513d15c261bf7ea7f6ae))

## [2.1.0](https://github.com/hrdtbs/noavoice/compare/v2.0.1...v2.1.0) (2020-03-05)


### Features

* add InstallPWA button ([#3](https://github.com/hrdtbs/noavoice/issues/3)) ([0360fec](https://github.com/hrdtbs/noavoice/commit/0360feca3021f05484a7e0d28f430ecee40a5b84))


### Bug Fixes

* **config:** disable audio cache ([aff71b7](https://github.com/hrdtbs/noavoice/commit/aff71b76c98f6d4e6ebdc0290920f16a21c7e6f3))
* **voice:** reduce processing weight ([ce6ec07](https://github.com/hrdtbs/noavoice/commit/ce6ec0715cc237aef15e441fcdaafd53e98e2d03))
* modify license link for mobile UI ([feb6e5a](https://github.com/hrdtbs/noavoice/commit/feb6e5a18bc3e6a205c7b3141c0c8bd7c1b713cb))
* move meta elements from _document to _app ([9daaa49](https://github.com/hrdtbs/noavoice/commit/9daaa49a9788492446b6637b24c0f685120b2eb6))
* remove unnecessary description ([ecd4702](https://github.com/hrdtbs/noavoice/commit/ecd4702f2df2beba881b8ac6bdca7758c1b7f3d5))
* upgrade to zero config for zeit now ([badf653](https://github.com/hrdtbs/noavoice/commit/badf6533126794ce0354a77a5a58bfa6b93efa0b))

### [2.0.1](https://github.com/hrdtbs/noavoice/compare/v2.0.0...v2.0.1) (2020-03-01)


### Bug Fixes

* add support of PWA Optimized ([7e28dca](https://github.com/hrdtbs/noavoice/commit/7e28dca416679e1084d8a5134011a3061a3e472d))
* make global style to native ([4d1f518](https://github.com/hrdtbs/noavoice/commit/4d1f518fe198cf30eb3a5104bd358d21a5183b24))

## [2.0.0](https://github.com/hrdtbs/noavoice/compare/v1.5.0...v2.0.0) (2020-02-29)


### Features

* use workbox ([67e3a2c](https://github.com/hrdtbs/noavoice/commit/67e3a2c8b4753586ed22bbb4c24716a1aeb1c57a))


### Bug Fixes

* follow nextjs changes ([2a68e6e](https://github.com/hrdtbs/noavoice/commit/2a68e6e66fbf7aa9f36e0a8e7f5e85bc2f5cdeae))

## [1.5.0](https://github.com/hrdtbs/noavoice/compare/v1.4.0...v1.5.0) (2020-02-29)


### Features

* add duration animation to audio ([010cc9c](https://github.com/hrdtbs/noavoice/commit/010cc9cb98e0b98c85f3a698bcb26b637ef70b1c))
* preconnect Google fonts ([998032b](https://github.com/hrdtbs/noavoice/commit/998032b58dfa3bafb923d6e16bb440ffaa1f7d24))
* use webp to image ([93ccd17](https://github.com/hrdtbs/noavoice/commit/93ccd1789caa657caccd33e9950856e565cd8a89))


### Bug Fixes

* add alt to twitter image ([408169b](https://github.com/hrdtbs/noavoice/commit/408169b1e76198cb06e791bcfe5b2afbe271aba1))
* compress header image ([b4b7478](https://github.com/hrdtbs/noavoice/commit/b4b7478291eaaabb1679e0ce754c70a6dabf685f))
* disable Google font ([c900cfb](https://github.com/hrdtbs/noavoice/commit/c900cfbccf170123b1a665220ae726ff77e1ef05))
* preconnect Google analytics ([4410c65](https://github.com/hrdtbs/noavoice/commit/4410c65829f2dfd633eced67fd1c09f2d29ae103))
* remove invalid audio ([958f052](https://github.com/hrdtbs/noavoice/commit/958f0527e1cefb0f350c6a075fc4d75bd031f5e1))
* remove unused CSS ([582dcd3](https://github.com/hrdtbs/noavoice/commit/582dcd32dc74dc9fe017263dd0823b078801ec42))

## [1.4.0](https://github.com/hrdtbs/noavoice/compare/v1.3.0...v1.4.0) (2020-02-29)


### Features

* add new audio files ([8410c6b](https://github.com/hrdtbs/noavoice/commit/8410c6b2fd03981d814ee9dddfcff181b1535ce3))
* add share button ([1613255](https://github.com/hrdtbs/noavoice/commit/1613255e1366b15ec680bab7f897fd88a41b5dc4))
* change font to Kosugi Maru ([3dac6ec](https://github.com/hrdtbs/noavoice/commit/3dac6ecf0514ec408ee1c6cbedb37d61fb749adc))
* follow audio src to change audio file extension ([81fc36d](https://github.com/hrdtbs/noavoice/commit/81fc36de0e8453d42b8cf2351fc670de4a968285))
* make voicelist ([15231e8](https://github.com/hrdtbs/noavoice/commit/15231e8c1cecd2a497ec0ef38a4c56ba247f2aca))


### Bug Fixes

* make audio preload to none ([7e57434](https://github.com/hrdtbs/noavoice/commit/7e57434b84275d0883ed9fbd277f643e58fe5305))

## [1.3.0](https://github.com/hrdtbs/noavoice/compare/v1.1.0...v1.3.0) (2020-02-26)


### Features

* add ogp settings ([fd0dede](https://github.com/hrdtbs/noavoice/commit/fd0dedee959a2f32112d3b7f7582af396915838a))
* add sitemap ([e34b33b](https://github.com/hrdtbs/noavoice/commit/e34b33be322fb655e81d25e89757fe60c84cc3c7))
* **pages/:** faviconの追加 ([ac12ebc](https://github.com/hrdtbs/noavoice/commit/ac12ebcebc18cb762fd708b057f58f09547cf09e))


### Bug Fixes

* change twitter ogp image ([bab7c20](https://github.com/hrdtbs/noavoice/commit/bab7c20941b7ba314cf3dac14c846a5fbb1ecac0))
* encode share path ([b45f9af](https://github.com/hrdtbs/noavoice/commit/b45f9af79609c297f8d1b4d2b8445dccbdd80d9b))
* lint:style --fix ([22e4315](https://github.com/hrdtbs/noavoice/commit/22e431557d6bafcc97bee49dd8a36bdef2d2e8f6))
* ogp ([8cb4e32](https://github.com/hrdtbs/noavoice/commit/8cb4e32a3bded3c3bc14fded05c22931fa61a408))
* update ([ca501b6](https://github.com/hrdtbs/noavoice/commit/ca501b6dee3c8a2bc1ec1b8ff4b5be40328727a1))
* update ([2c473a0](https://github.com/hrdtbs/noavoice/commit/2c473a0517b2213e07f865c1a3fd37f8298a9991))
* **pages/:**  ヘッダーイメージのサイズを軽量化した ([d1625e3](https://github.com/hrdtbs/noavoice/commit/d1625e33ef5f516ef934462ac283def054d3bf25))
* **pages/:** make image decoding to  async ([f69fd36](https://github.com/hrdtbs/noavoice/commit/f69fd36b17a9701ea7246d5c14fee128489622c7))

## [1.2.0](https://github.com/hrdtbs/noavoice/compare/v1.1.0...v1.2.0) (2020-02-03)



pages/index.js
import { useCallback, useState } from "react"
import { voices } from "../constants/voices"
import clsx from "clsx"

const sx = {
    title: {
        textAlign: "right",
        fontSize: "25px",
        fontFamily: "serif"
    },
    voiceGroup: {}
}

const Voice = ({ name }) => {
    const [el, setEl] = useState(null)
    const [duration, setDuration] = useState(0)
    const [active, setActive] = useState(false)
    const handleClick = useCallback(() => {
        if (el) {
            setDuration(el.duration)
            el.play()
        }
    }, [el])

    return (
        <figure onClick={handleClick} className={clsx(active && "active")}>
            <style jsx>{`
                figure {
                    position: relative;
                    display: inline-flex;
                    flex: 1 1 auto;
                    justify-content: space-between;
                    padding: 12px 20px 12px 32px;
                    margin: 6px;
                    overflow: hidden;
                    color: white;
                    cursor: pointer;
                    background-color: #faa65f;
                    border-left: 12px solid #fcdfa1;
                    border-radius: 10px;
                    box-shadow: 0 3px 1px -2px rgba(0, 0, 0, 0.2), 0 2px 2px 0 rgba(0, 0, 0, 0.14),
                        0 1px 5px 0 rgba(0, 0, 0, 0.12);
                }
                figure::before {
                    content: "";
                    display: block;
                    position: absolute;
                    top: 0px;
                    left: 0px;
                    height: 100%;
                    background-color: black;
                    mix-blend-mode: soft-light;
                    will-change: width;
                    transform: translate3d(0, 0, 0);
                }
            `}</style>
            <style jsx>{`
                figure.active::before {
                    animation: decrement ${duration}s ease-in-out;
                }
                @keyframes decrement {
                    0% {
                        width: 100%;
                    }
                    100% {
                        width: 0%;
                    }
                }
            `}</style>
            <figcaption>{name}</figcaption>
            <audio
                preload="metadata"
                ref={node => {
                    setEl(node)
                }}
                src={`/static/voices/${name}.mp4`}
                onPlay={() => setActive(true)}
                onEnded={() => setActive(false)}
            ></audio>
        </figure>
    )
}

function IndexPage() {
    return (
        <main>
            <style jsx>{`
                main {
                    max-width: 900px;
                    margin: auto;
                    box-shadow: 0 3px 1px -2px rgba(0, 0, 0, 0.2), 0 2px 2px 0 rgba(0, 0, 0, 0.14),
                        0 1px 5px 0 rgba(0, 0, 0, 0.12);
                    background-color: #fff;
                    padding-bottom: 24px;
                }
                main > header img {
                    object-fit: cover;
                }
                main > nav {
                    display: flex;
                    justify-content: space-between;
                    margin: 6px 32px;
                    flex-wrap: wrap;
                }
                main > article {
                    display: flex;
                    flex-wrap: wrap;
                    margin: 6px 12px;
                }
                a {
                    color: #ff7500;
                    text-decoration: none;
                    transition: 0.3s;
                }

                a:hover {
                    color: #faa65f;
                    text-decoration: underline;
                }
                a button {
                    all: unset;
                    display: flex;
                    align-items: center;
                    justify-content: center;
                    padding: 1em 0.5em;
                    margin: 0.5em 0px;
                    border: 4px dotted #faa65f;
                }
            `}</style>
            <header>
                <picture>
                    <source srcSet="/static/header-min.webp" type="image/webp" />
                    <img src="/static/header-min.png" alt="header" height="150px" width="100%" loading="lazy" />
                </picture>
            </header>

            <nav>
                <div>
                    <small>
                        <a href="https://github.com/hrdtbs/noavoice/blob/master/CHANGELOG.md">noavoice: v2.0.0</a>
                        <br />
                        <a href="https://twitter.com/_noach" style={sx.descriptionButton}>
                            声：望月のあ @_noach
                        </a>
                        <br />
                        <a
                            rel="noreferrer noopener"
                            target="_blank"
                            href="https://github.com/hrdtbs/noavoice#%E9%9F%B3%E5%A3%B0%E3%83%95%E3%82%A1%E3%82%A4%E3%83%AB%E3%81%AElicence%E3%81%AB%E3%81%A4%E3%81%84%E3%81%A6"
                        >
                            音声ファイルのライセンスについて
                        </a>
                    </small>
                </div>
                <a
                    href={encodeURI(
                        `https://twitter.com/intent/tweet?url=https://noavoice.now.sh&text=Vtuber望月のあボイスボタン&hashtags=のあぼいす`
                    )}
                    target="_blank"
                    rel="noopener noreferrer"
                >
                    <button>
                        <img src="/static/twitter-blue.svg" width="32px" height="32px" alt="twitter" />
                        Twitterでシェアして応援！
                    </button>
                </a>
            </nav>
            <article>
                {voices.map((voice, i) => {
                    return <Voice name={voice} key={`voice-${i}`} />
                })}
            </article>
        </main>
    )
}

export default IndexPage
